[#187364738] - A seller should be able to add/create a product

Furaha-Justine opened the pull request:
**PR Description**

**Summary:**
This pull request introduces a crucial feature allowing sellers to seamlessly add new products with associated images to our application.

**Changes Made:**

- **Endpoint Addition**: Implemented a new endpoint /api/products/add dedicated to handling the addition of products to our system.
- **Controller Logic:** Developed the addProduct controller function responsible for orchestrating the addition of products, ensuring efficient and organized handling of incoming requests.
- **Image Upload Integration**: Integrated Cloudinary, a cloud-based image management solution, to facilitate secure storage and management of product images. This integration ensures reliable image handling without burdening our local server.
- **Validation Enhancements:** Implemented robust validation checks within the controller logic to ensure the integrity and consistency of the data being added. This includes checks for required fields and prevention of duplicate product names for the same seller, fostering data accuracy and reliability.
- **API Documentation:** Enriched the codebase with Swagger documentation for the addProduct endpoint, enhancing the clarity and accessibility of our API documentation. This ensures that developers can easily understand and utilize this new functionality.

**How to Test:**

- **Start Application:** Begin by launching the application environment.
- **API Request:** Send a POST request to the newly added /product endpoint, providing valid product details and attaching corresponding images.
- **Validation Checks:** Ensure that the request is processed successfully, validating that the product is added to the database and the images are securely uploaded to Cloudinary.
- **Error Handling:** Validate the system's behavior by attempting to add a product with the same name for the same seller, confirming that appropriate error messages are returned.
- **Data Integrity Testing:** Test the endpoint with various combinations of valid and invalid data inputs to verify the system's robustness and resilience to different scenarios.

**Link**
https://www.loom.com/share/3cca87b5618f47bba8d1532d11a2e962?sid=c542e344-9f1b-49b0-ae7f-9bcab89b4b4f

**Additional Notes:**

**Configuration Requirements:** Before testing, ensure that the required environment variables for Cloudinary (CLOUDINARY_CLOUD_NAME, CLOUDINARY_API_KEY, CLOUDINARY_API_SECRET) are correctly configured to enable seamless image handling.

Pivotal Tracker Story
#187364738

EricNiyo73 reviewed (commented):
Please add a video or screenshot demonstrating how this feature works.

AngeloChristian1 commented on src/controllers/LoginController.ts:
```diff
@@ -11,6 +11,13 @@ import sendMail from '../utils/sendEmail'
+   * @param {string} role - The role of the user to include in the token payload
+   * @returns {string} JWT token
+   */
+
```
What are these for?

AngeloChristian1 commented on src/controllers/productController.ts:
```diff
@@ -0,0 +1,66 @@
+  static async addProduct(req: Request, res: Response): Promise<Response> {
+    try {
+      const { name, price, categoryId, expiryDate, bonus, sku } = req.body
+      if (!name || !price || !categoryId || !expiryDate || !bonus || !sku) {
```
I think it is more convenient for a product to have a category name no categoryId, but even this is ok

AngeloChristian1 reviewed (commented):
You did a good job, but we can work on the following
- Creating a stock table for storing the Stock keeping unit (stu) for every product
- Rebasing properly so that you only have you files, you can achieve this by reseting to the most recent commit made on develop and the discarding files you didn't modify.

So can work on those changes.
Regards

EricNiyo73 commented on src/controllers/productController.ts:
```diff
@@ -0,0 +1,118 @@
+  ): Promise<Response> {
+    try {
+      const { name } = req.body
+      if (!name) {
+        return res.status(400).json({ message: 'Bad request' })
+      }
```
It seems kike like you are repeating some snippets of codes, I think this error handling must be done by validateRrequest middleware , which means you  have to pass it into Product route.

```
router.post(
  '/collection',
  isAuthenticated, validateRequest(CollectionSchema, 'body'),
  checkPermission(UserRole.SELLER),
  productController.createCollection,
)
```

EricNiyo73 commented on src/controllers/productController.ts:
```diff
@@ -0,0 +1,118 @@
+        quantity,
+        expiryDate,
+      } = req.body
+      if (
+        !name ||
+        !price ||
+        !category ||
+        !collectionId ||
+        !bonus ||
+        !sku ||
+        !quantity ||
+        !expiryDate
+      ) {
+        return res.status(400).json({ message: 'Bad request' })
+      }
```
you are repeating some snippets of codes, I think must be done by joi schema  validation, You have passed out into your route.

EricNiyo73 reviewed (changes requested):
It seems like you are repeating some codes, just try to handle that duplicates.

EricNiyo73 commented on src/controllers/productController.ts:
```diff
@@ -0,0 +1,103 @@
+    req: Request,
+    res: Response,
+  ): Promise<Response> {
+    try {
+      const { name } = req.body
+
+      const collection = await Collection.create({ name })
```
Also here you can save the user (userId) who created that collection, this can be helpful.

Furaha-Justine committed:
Feature: Add product on seller side

modify file

modify product controller and its tests

changes routes

update tests

update tests

munezerobagira commented on src/routes/productRoute.ts:
```diff
@@ -0,0 +1,28 @@
+  configureMulter,
+  validateRequest(productSchema, 'body'),
+  productController.addProduct,
+)
```
```suggestion
router.post(
  '/collections/:collectionId/items',
  isAuthenticated,
  checkPermission(UserRole.SELLER),
  configureMulter,
  validateRequest(productSchema, 'body'),
  productController.addProduct,
)
```
it would better if the route is descriptive

munezerobagira commented on src/database/models/collectionModel.ts:
```diff
@@ -0,0 +1,62 @@
+          type: DataTypes.STRING,
+          allowNull: false,
+          unique: true,
+        },
```
We can also add description  for the collection

munezerobagira commented on src/controllers/productController.ts:
```diff
@@ -0,0 +1,107 @@
+        name,
+        price,
+        category,
+        collectionId,
```
You need to check if the seller is adding product to collection that exists
